### PR TITLE
update rom file search order

### PR DESCRIPTION
**Pull request by n-at**
Update to ROM file search algorithm.
Prevents cue file creation if single-file image (like iso, cso, ...) is loaded or cue file (or ccd, m3u, ...) already exists.
If none found, creates cue (if core supports it) or loads any file with supported extension.

### CODE REVIEW

**Comment by ethanaobrien** [comment]
If I remember correctly, the reason why I made it always generate a cue file is because on the mednafen cores (at least the Playstation one) its case sensitive and path sensitive and as a result is a common user error. Have you checked this?

**Comment by n-at** [comment]
I checked yabause (segaSaturn) - there was issues with generated cue files, this algorithm solved it.

Now checked:
* pcsx_rearmed - works
* mednafen_psx_hw - works
File names inside archive was:
```
Metal Gear Solid (USA) (Disc 1)
├── Metal Gear Solid (USA) (Disc 1).bin
└── Metal Gear Solid (USA) (Disc 1).cue
```
(maybe change file names to something else to test?)
* mednafen_pcfx - works

File: data/emulator.js [hunk 1]
```diff
@@ -890,49 +900,58 @@ class EmulatorJS {
...
-                        if (k === "!!notCompressedData") {
-                            if (needsCue && execFile !== null) {
-                                this.fileName = execFile;
+                    let isoFile = null;
+                    let supportedFile = null;
+                    let cueFile = null;
+                    let selectedCueExt = null;
+                    fileNames.forEach(fileName => {
+                        const ext = fileName.split('.').pop().toLowerCase();
+                        if (supportsExt(ext)) {
```
Comment (ethanaobrien) [inline data/emulator.js]: Could we only set this if it's not already defined? (Set it to the first file it finds)

File: data/emulator.js [hunk 2]
```diff
@@ -890,49 +900,58 @@ class EmulatorJS {
...
+                            supportedFile = fileName;
+                        }
+                        if (['iso', 'cso', 'chd', 'elf'].includes(ext)) {
+                            isoFile = fileName;
+                        }
+                        if (['cue', 'ccd', 'toc', 'm3u'].includes(ext)) {
+                            if (this.getCore(true) === 'psx') {
+                                //always prefer m3u files for psx cores
+                                if (selectedCueExt !== 'm3u') {
+                                    if (ext === 'm3u' || cueFile === null) {
```
Comment (ethanaobrien) [inline data/emulator.js]: Can we also only set this if it hasn't been set

File: data/emulator.js [hunk 3]
```diff
@@ -890,49 +900,58 @@ class EmulatorJS {
...
+                                        cueFile = fileName;
+                                        selectedCueExt = ext;
+                                    }
+                                }
                             } else {
-                                this.fileName = altName;
+                                //prefer cue or ccd files over toc or m3u
+                                if (!['cue', 'ccd'].includes(selectedCueExt)) {
+                                    if (['cue', 'ccd'].includes(ext) || cueFile === null) {
```
Comment (ethanaobrien) [inline data/emulator.js]: And same here

File: data/emulator.js [hunk 4]
```diff
@@ -890,49 +900,58 @@ class EmulatorJS {
...
-                            //always prefer m3u files for psx cores
-                            !(this.getCore(true) === "psx" && ["m3u", "ccd"].includes(this.fileName.split(".").pop())))) {
-                            this.fileName = k;
-                        }
-                        if (needsCue && execFile === null && ["m3u", "cue"].includes(k.split(".").pop().toLowerCase())) {
-                            FS.writeFile("/"+k, resData[k]);
+                    });
+                    if (isoFile !== null && (supportsExt('iso') || supportsExt('cso') || supportsExt('chd') || supportsExt('elf'))) {
+                        this.fileName = isoFile;
+                    } else if (!this.config.disableCue && (supportsExt('cue') || supportsExt('ccd') || supportsExt('toc') || supportsExt('m3u'))) {
```
Comment (ethanaobrien) [inline data/emulator.js]: On the pcsx reamed core, I think we should enable `disableCue` by default

**Review by ethanaobrien — CHANGES_REQUESTED** [review]

**Comment by ethanaobrien** [comment]
Hmm, alright. I left a review. Once those things are addressed I'll merge

**Comment by n-at** [comment]
Added changes. 

The only problem with file name case I found, if file structure is like this:
```
METAL Gear Solid (USA) (Disc 1)
├── Metal GEAR Solid (USA) (Disc 1).bin
└── Metal Gear SOLID (USA) (Disc 1).cue
```
and the cue file is:
```
FILE "Metal Gear Solid (USA) (Disc 1).bin" BINARY
  TRACK 01 MODE2/2352
    INDEX 01 00:00:00
```
it will not load. I don't know if it's an issue. This file structure will only work on Windows, Linux/Mac will also not be able to load the file.

File: data/emulator.js (hunk 2 above)
Comment (n-at) [reply data/emulator.js]: It's already doing this:
* if matching file found, takes first
* if m3u file found, takes first and replaces current selected (it it wasn't already m3u)

File: data/emulator.js (hunk 4 above)
Comment (n-at) [reply data/emulator.js]: Also enabled for genesis_plus_gx, picodrive and mednafen_pce. They can load both cue and single-file roms.

**Comment by ethanaobrien** [comment]
>it will not load. I don't know if it's an issue. This file structure will only work on Windows, Linux/Mac will also not be able to load the file.

Not technically. The emscripten fs is case sensitive

**Pull request merged**